Gale Databases.js

```diff
@@ -25,93 +26,222 @@
-	switch (entries.length) {
-		case 0: break;
-		case 1: processSingleEntry(doc); break;
-		default: processMultipleEntries(entries);
+function doWeb(doc, url) {
+	if (detectWeb(doc, url) == "multiple") {
+		Zotero.selectItems(getSearchResults(doc, false), function (items) {
+			if (items) ZU.processDocuments(Object.keys(items), scrape);
+		});
+	}
+	else {
+		scrape(doc, url);
 	}
 }
 
+function scrape(doc, url) {
+	let citeData = doc.querySelector('input.citationToolsData');
+	let documentUrl = decodeURIComponent(citeData.dataset.url);
+	let mcode = citeData.dataset.mcode; // undefined for bookSections
+	let productName = citeData.dataset.productname;
+	let docId = mcode ? undefined : decodeURIComponent(url.match(/(?:docId|id)=([^&]+)/)[1]);
+	let risPostBody = JSON.stringify([{
+		documentUrl: `<span class="docUrl">${documentUrl}</span>`,
+		mcode,
+		docId,
+		productName
+	}]);
+	let pdfURL = attr(doc, 'button[data-gtm-feature="download"]', 'data-url');
 
+	ZU.doPost('https://go.gale.com/ps/citationtools/rest/cite/getcitations/', risPostBody, function (text) {
+		let citations = JSON.parse(text);
+		let translator = Zotero.loadTranslator("import");
+		translator.setTranslator("32d59d2d-b65a-4da4-b0a3-bdd3cfb979e7"); // RIS
+		translator.setString(citations.RIS[0]);
+		translator.setHandler("itemDone", function (obj, item) {
+			if (pdfURL) {
+				item.attachments.push({
+					url: pdfURL,
+					title: "Full Text PDF",
+					mimeType: "application/pdf"
+				});
+			}
+			item.attachments.push({
+				title: "Snapshot",
+				document: doc
+			});
+			item.notes = [];
+			item.complete();
+		});
+		translator.translate();
+	}, {
+		'Content-Type': 'application/json; charset=utf-8',
+		'Accept': 'application/json'
+	});
+}
 /** BEGIN TEST CASES **/
 var testCases = [
+	{
+		"type": "web",
+		"url": "https://go.gale.com/ps/eToc.do?contentModuleId=GVRL&resultClickType=AboutThisPublication&searchType=BasicSearchForm&docId=GALE%7C5BBU&userGroupName=ucberkeley&inPS=true&rcDocId=GALE%7CCX3427400124&prodId=GVRL",
+		"items": [
+			{
+				"itemType": "book",
+				"title": "Arts and Humanities Through the Eras",
+				"creators": [
+					{
+						"lastName": "Bleiberg",
+						"firstName": "Edward I.",
+						"creatorType": "editor"
+					},
+					{
+						"lastName": "Evans",
+						"firstName": "James Allan",
+						"creatorType": "editor"
+					},
+					{
+						"lastName": "Figg",
+						"firstName": "Kristen Mossler",
+						"creatorType": "editor"
+					},
+					{
+						"lastName": "Soergel",
+						"firstName": "Philip M.",
+						"creatorType": "editor"
+					},
+					{
+						"lastName": "Friedman",
+						"firstName": "John Block",
+						"creatorType": "editor"
+					}
+				],
+				"date": "2005",
+				"archive": "Gale eBooks",
+				"libraryCatalog": "Gale",
+				"place": "Detroit, MI",
+				"publisher": "Gale",
+				"series": "Ancient Egypt 2675-332 B.C.E.",
+				"url": "https://link.gale.com/apps/pub/5BBU/GVRL?u=ucberkeley&sid=GVRL",
```

Comment:
You said if you remove the `u=` you get a login page. Did you mean a Gale login page or a Berkeley proxy login page?

I know the old translator did this, but it doesn't seem like it makes sense to includes these in citations if they're not usable to others, so we'd only want to save a URL that someone at a different institution could actually use.

(I don't currently have access to test this.)

Reply:
A Gale login page. It doesn't seem to matter whether the `u=` value matches the institution you're connecting from; I can access the link in the last test with `u=nysl_ce_syr`. (I need to hunt down some links from other institutions to make sure it isn't just that one.) And the links the translator is saving are the same ones that the `Get Link` button on the page gives you, which are supposedly permalinks that you can cite/share.

Reply:
Hmm, now I see that the tests with Berkeley URLs fail on CI but the New York URL doesn't. Odd. I'll find more cases to test with.